### ISSUE TITLE
Introduce TypeAssert to constraint checker to certain type with a proxy check

### ISSUE THREAD

**Issue by mahmoudajawad** [python/mypy_extensions#45]
This is a follow-up to https://github.com/python/mypy/issues/14846.

The suggestion is to add `TypeAssert` generic typing constraint utility that enforces checker to cast type per the value. This is a suggestion by @erictraut, and he explained it in https://mail.python.org/archives/list/typing-sig@python.org/thread/JTO3WKRKQFW5YFP3ZSMNTRIEA5NPJEM6/ . He supplied following example as well:

```python
def verify_no_nones(val: list[None | T]) -> TypeAssert[list[T]]:
    if None in val:
        raise ValueError()

def func(x: list[int | None]):
    verify_no_nones(x)
    reveal_type(x)  # list[int]

def assert_is_one_dimensional(val: tuple[T, ...] | T) -> StrictTypeAssert[tuple[T] | T]:
    if isinstance(val, tuple) and len(val) != 1:
        raise ValueError()

def func(x: float, y: tuple[float, ...]):
    assert_is_one_dimensional(x)
    reveal_type(x)  # float

assert_is_one_dimensional(y)
    reveal_type(y)  # tuple[float]
```

I'm growing in need of this functionality by the day, and suggestion of @erictraut seems to be the solution that works perfect for me as well. I want to request getting this accepted to `mypy_extensions`, that is maintainers giving green-light to work on such a feature. I would be more than happy to work on this as I'm the person who is in need of it. I might need guidance though.

Do I have the OK to work on this?

**Comment by erictraut** [comment]
The [python/typing discussion forum is probably a better place to initiate this discussion](https://github.com/python/typing/discussions). I had never even heard of this forum before and wouldn't have noticed your post if you hadn't at-mentioned me. The other forum includes most of the active typing community members, including the maintainers of typeshed and all the major static and runtime type checkers.

[Here](https://github.com/python/typing/discussions/1013) is the thread on the python/typing forum where I first proposed `TypeAssert`.

A couple of additional thoughts...

There is precedent for this functionality in TypeScript, where it's referred to as [type assertions](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions). TypeScript even has dedicated syntax for the feature. It might be worth doing a bit of digging to see if any other modern languages offer a similar feature. Conducting this sort of survey can uncover new ideas or identify potential pitfalls early in the design process. I did something like this for [PEP 695](https://peps.python.org/pep-0695/#appendix-a-survey-of-type-parameter-syntax), and it proved to be very useful.

Rich Chiodo (@rchiodo) is [working on a draft PEP](https://github.com/python/typing/discussions/1013#discussioncomment-6579814) that covers the related topic of "strict type guards". It might make sense to combine the two topics into one. I'd look to Rich and @JelleZijlstra, the PEP's sponsor, to offer guidance on the pros/cons of combining the two into a single PEP. If the decision is made to keep them separate, I encourage you to still coordinate with Rich. In that case, you'll need to follow the [documented PEP workflow](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) and find a sponsor.

**Comment by JelleZijlstra** [comment]
I agree. We're unlikely to implement this feature in mypy-extensions in any case: as Eric says the preferred flow would be writing a PEP, then adding it to typing-extensions. mypy-extensions at this point is only for truly mypy-specific extensions, such as things that make sense only with mypyc.
